Add Accept & Content-Type headers to rest api spec

Mpdreamz opened the pull request:
This documents the preferred `Accept` and `Content-Type` headers to be
sent with API calls.


Elasticsearch supports more content types e.g `CBOR` and `smile` but none of
our REST clients actively support these so we don't make an active
effort documenting these.

In some cases there `response_accepts` documents multiple formats e.g for the `_cat` API's to signal the preferred mimetype is `text/plain` but still advertise the fact it can return `application/json`.

Mpdreamz committed:
Add Accept & Content-Type headers to rest api spec

This documents the preferred Accept and Content Type headers to be
sent with API calls.

In some cases the API returns two formats e.g cat API's in which case
the `headers.accept` documents both in order of preference.

Elasticsearch supports more content types e.g CBOR and smile but none of
our REST clients actively support these so we don't make an active
effort documenting these.

rjernst commented:
@Mpdreamz Could we perhaps keep these two types separate (accepted input to the api, and the output type of the api)?

I'm also not seeing where cat apis can return json, at least not in the code. Could you clarify this point or show an example of one of our apis that does this?

Mpdreamz commented:
Hey @rjernst 

The two are separate `response_accepts` is what ES returns and `content_type` on `body` is what ES accepts. 

The cat API can return json, the .NET high level client uses this (although all low level clients including .NET default to text/plain for the cat APIs). 

See: https://github.com/elastic/elasticsearch/blob/99f88f15c5febbca2d13b5b5fda27b844153bf1a/server/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java#L61

rjernst commented:
Thanks for the pointer. I was looking at AbstractRestAction, which has the "help" response that seems to be hardcoded to text. I also misread the diff; I see now the separation of content-type.

Content-Type matches what the http spec defines for the request type. In the spec the response is simply with the "Accept" header, which is obviously too generic here to be meaningful, so I see why you have gone with "response_accepts". However, since we aren't matching the spec exactly anyways, could we go with something a little more clear like "allowed_response_types" or just "response_types" or something like that?

dakrone commented:
All of our APIs also support outputting YAML, should those be added to the list? What about CBOR and SMILE (which we also support)?

Mpdreamz commented:
@rjernst  I like `response_types` will roll with that :+1:

@dakrone since none of the clients actively support these (other than being able to override accept/content-type) headers I opted to not document these. Also because their usefulness/existence has been questioned multiple times in the past.

russcam commented:
Since `content_type` is used for Content-Type, would `accept` be better for Accept than `response_types`? 

Thinking about this more, introducing a `headers` object for these properties might be a good idea; we might want to document where `x-opaque-id` can be used in future, as well as security related headers.

Mpdreamz commented:
`headers` was my first implemention but then switched to this since it would be easier to add validation to make `content_type` required on the basis of body being present in the current parser. 

Since we're now exploring doing more rigurous validation I'm going to switch back to `headers` since it's more extensible and self descriptive :+1:

Mpdreamz committed:
update unit tests

Mpdreamz committed:
fix find replace in funny-timeout-watch.json

Mpdreamz commented:
@elasticmachine run elasticsearch-ci/2

Mpdreamz commented:
@russcam I've rebased this against your changes that have landed in `master` and reworked this to move to `headers:{}` + updated the json schema accordingly.

russcam commented on rest-api-spec/src/main/resources/rest-api-spec/api/cat.help.json:
```diff
@@ -5,6 +5,9 @@
     },
     "stability":"stable",
+    "headers":{
+      "accept": [ "text/plain", "application/json"]
```
```suggestion
      "accept": [ "text/plain" ]
```

Cat help looks like it only supports `text/plain`

russcam commented on rest-api-spec/src/main/resources/rest-api-spec/api/msearch_template.json:
```diff
@@ -5,6 +5,10 @@
     "stability":"stable",
+    "headers":{
+      "accept": [ "application/json"],
+      "content_type": ["application/x-ndjson"]
```
Since this technically supports sending `application/json` too, should we add it after the preferred `application/x-ndjson`? I can see an argument against not including it as even though the `content-type` header is `application/json`, the request body format must be newline delimited JSON

russcam commented on rest-api-spec/src/main/resources/rest-api-spec/api/nodes.reload_secure_settings.json:
```diff
@@ -5,6 +5,9 @@
     },
     "stability":"stable",
+    "headers":{
+      "accept": [ "application/json"]
```
```suggestion
      "accept": [ "application/json"],
      "content_type": [ "application/json" ]
```

Takes a body: https://www.elastic.co/guide/en/elasticsearch/reference/7.x/cluster-nodes-reload-secure-settings.html

russcam commented on rest-api-spec/src/main/resources/rest-api-spec/api/search_template.json:
```diff
@@ -100,8 +104,7 @@
     "body":{
-      "description":"The search definition template and its params",
-      "required":true
+      "description":"The search definition template and its params"
```
intentional to remove `required: true`?

Mpdreamz replied:
Thank you for catching this one 👍

russcam commented on rest-api-spec/src/main/resources/schema.json:
```diff
@@ -142,7 +145,30 @@
+            },
+            "required": ["accept"],
+            "title": "Headers the API understands",
+            "description": "Documents the headers the API supports, at a minimum this should document the Accept header and Content-Type header if the API supports a body"
```
```suggestion
            "description": "Documents the headers the API supports, at a minimum this must document the Accept header. The Content-Type header should be documented if the API supports a body"
```

russcam commented on x-pack/plugin/src/test/resources/rest-api-spec/api/ml.post_data.json:
```diff
@@ -5,6 +5,10 @@
     "stability":"stable",
+    "headers":{
+      "accept": [ "application/json"],
+      "content_type": ["application/x-ndjson", "application/json"]
```
Similar to other APIs that expect newline delimited JSON request body - I think we should be consistent on whether we document `application/json` as accepted Content-Type across these APIs

russcam commented on x-pack/qa/full-cluster-restart/src/test/resources/org/elasticsearch/xpack/restart/funny-timeout-watch.json:
```diff
@@ -35,4 +35,4 @@
   }
-}
\ No newline at end of file
+}
```
guess this change can be discarded?

Mpdreamz replied:
reverting it back is hard inside the IDE, because the old invalidates editorconfig. I have to close the editor and edit it in vim to stop it automatically reverting back to the state in the PR. Leaving as is but good catch again!

russcam commented on rest-api-spec/src/main/resources/rest-api-spec/api/msearch.json:
```diff
@@ -5,6 +5,10 @@
     "stability":"stable",
+    "headers":{
+      "accept": [ "application/json"],
+      "content_type": ["application/x-ndjson"]
```
Since this technically supports sending `application/json` too, should we add it after the preferred `application/x-ndjson`? I can see an argument against not including it as even though the `content-type` header is `application/json`, the request body format must be newline delimited JSON

russcam commented on x-pack/plugin/src/test/resources/rest-api-spec/api/ml.find_file_structure.json:
```diff
@@ -5,6 +5,10 @@
     "stability":"experimental",
+    "headers":{
+      "accept": [ "application/json"],
+      "content_type": ["application/x-ndjson"]
```
Since this technically supports sending `application/json` too, should we add it after the preferred `application/x-ndjson`? I can see an argument against not including it as even though the `content-type` header is `application/json`, the request body format must be newline delimited JSON

russcam reviewed (changes requested):
I left some comments.

My thought about APIs accepting newline-delimited JSON bodies is that we should also document that these APIs will also accept a `content_type` of `application/json`, even though the body is newline delimited. The specs would then match the behaviour.

Mpdreamz committed:
Update rest-api-spec/src/main/resources/rest-api-spec/api/cat.help.json

Co-authored-by: Russ Cam <russ.cam@elastic.co>
(cherry picked from commit a3b7283df310e61ba5a71fa1e3b788dfbec59420)

Mpdreamz commented on rest-api-spec/src/main/resources/rest-api-spec/api/bulk.json:
```diff
@@ -5,6 +5,10 @@
     "stability":"stable",
+    "headers":{
+      "accept": [ "application/json"],
+      "content_type": ["application/x-ndjson"]
```
Aye that's why I opted not to include it. This documents the preferred mimetypes, to be extended with API versioning headers later on as well. The fact the server is lenient here is not a feature for clients IMO.

Mpdreamz committed:
Update rest-api-spec/src/main/resources/rest-api-spec/api/nodes.reload_secure_settings.json

Co-authored-by: Russ Cam <russ.cam@elastic.co>
(cherry picked from commit 4f0cf97157b795f3a476bd57df16fddfd72913fe)

Mpdreamz committed:
Update rest-api-spec/src/main/resources/schema.json

Co-authored-by: Russ Cam <russ.cam@elastic.co>
(cherry picked from commit 1f872d5429c21f05e0f9d23fcc5fd3ef8703766e)

Mpdreamz committed:
body required on search_template

Mpdreamz committed:
add headers to new API

Mpdreamz committed:
add headers to snapshot.clone

Mpdreamz committed:
validate xpack API's

Mpdreamz commented:
jenkins run elasticsearch-ci/2

Mpdreamz commented:
> I left some comments.
> 
> My thought about APIs accepting newline-delimited JSON bodies is that we should also document that these APIs will also accept a `content_type` of `application/json`, even though the body is newline delimited. The specs would then match the behaviour.

This PR is not meant to be exhaustive but document the the headers clients should be sending. For similar reasons I opted out from listing cbor,smile etcetera.

pgomulka reviewed (approved):
it looks great. I think this is a good first step.
I am also tempted that maybe all possible accept/content-type headers could be included in the spec. Including versioned one (for the current version). But I think this should be addressed later.
Also I wonder if we should start using this in our testing - actually sending the header value specified in the spec. Again, this should be handled separately. 

LGTM thank you
